### PR TITLE
Add helper functions allowing lambdas to be used to modify junit matchers

### PR DESCRIPTION
**Pull request by thecoop**
Add junit matchers allowing other matchers to be transformed using a function. This can make checking properties of lists/arrays a lot more fluent using nested matchers, rather than declaratively checking individual items. These replace the custom `ElasticsearchMatchers` with more generic ones.

As a basic example, it allows you to turn this:

    assertThat(list.get(0).getName(), equalTo("foo"));
    assertThat(list.get(1).getName(), equalTo("bar"));
    assertThat(list.get(2).getName(), equalTo("quux"));

into this:

    assertThat(list, transformedItems(Item::getName, contains("foo", "bar", "quux")));

Doing this 'properly' without these helpers requires defining your own matchers, which is very cumbersome.

I've applied the new methods to `ElasticsearchAssertions` and a few other classes to show various use cases.

### CODE REVIEW

**Review by rjernst — COMMENTED** [review]
The concept sounds very useful! I left a few comments. In general I have 2 more thoughts:
* In the examples changed, there seem to be a few unrelated changes. I think these should be reverted and moved to a different PR, as they make this PR more difficult to see the actual change.
* I'm not fond of the name. "transform" does make sense in theory, but it looks more like a command/action method than a matcher method.

File: test/framework/src/main/java/org/elasticsearch/test/LambdaMatchers.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            this.transform = transform;
+        }
+
+        @Override
+        @SuppressWarnings("unchecked")
+        public boolean matches(Object actual) {
+            U u;
+            try {
+                u = transform.apply((T) actual);
+            } catch (ClassCastException e) {
```
[the remark this answers is not in the document]
Comment (thecoop) [reply test/framework/src/main/java/org/elasticsearch/test/LambdaMatchers.java:39]: ooh yes, I forgot about `not`

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java [hunk 2]
```diff
@@ -159,24 +164,14 @@ public static void assertBlocked(BaseBroadcastResponse replicatedBroadcastRespon
...
-                }
-                assertThat("Request should have been blocked by [" + expectedBlockId + "] instead of " + e.blocks(), found, equalTo(true));
-            }
+        var e = expectThrows(ClusterBlockException.class, builder::get);
+        assertThat(e.blocks(), not(empty()));
+        RestStatus status = checkRetryableBlock(e.blocks()) ? RestStatus.TOO_MANY_REQUESTS : RestStatus.FORBIDDEN;
+        assertThat(e.status(), equalTo(status));
+
+        if (expectedBlockId != null) {
+            boolean exists = e.blocks().stream().anyMatch(b -> b.id() == expectedBlockId);
```
[the remark this answers is not in the document]
Comment (thecoop) [reply test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java:173]: Good spot

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java [hunk 3]
```diff
@@ -283,7 +259,7 @@ public static void assertHitCount(SearchResponse countResponse, long expectedHit
 
     public static void assertExists(GetResponse response) {
         String message = String.format(Locale.ROOT, "Expected %s/%s to exist, but does not", response.getIndex(), response.getId());
-        assertThat(message, response.isExists(), is(true));
```
[the remark this answers is not in the document]
Comment (thecoop) [reply test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java:286]: I did a general refactor of `ElasticsearchAssertions` while I'm here. But the `assertThat` for booleans does give better messages, so I'll change that back.

**Comment by thecoop** [comment]
I'm also not sure about the name, although `transformed` does kinda work for readability:

`assertThat(x, transformedItems(contains(1, 2, 3)))` - 'assert that x has transformed items 1 2, 3'
`assertThat(x, transformed(greaterThat(7)))` - 'assert that x transformed is greater than 7'

**Comment by rjernst** [comment]
The confusing thing to me on the name is "transformed into what?"

**Comment by rjernst** [comment]
I don't think transformedWith quite works, it still doesn't translate to an understandable statement.

Thinking about this like other matcher functions are named, how about:

```
matchesTransform - assert x matches transformation abc
hasTransformedItems - assert that x has items matching transformation abc
hasTransformedArrayItems - assert that x has array items matching transformation abc
```

**Comment by thecoop** [comment]
`hasTransformedItems` doesn't work, as that conflicts with `hasItems`, which means a different thing. It depends on what the inner matcher actually does.

How about `[array]itemsMatchTransform` or `transformed[Array]ItemsMatch`?

**Comment by rjernst** [comment]
I'm in favor of using the "match(es)" suffix since the functions take a Matcher.

So `transformed[Array]ItemsMatch` from your list. For non items I guess it would be `transformedMatch`?

**Comment by thecoop** [comment]
Those work - renamed

**Review by rjernst — APPROVED** [review]
LGTM. I'm happy with where we ended up, the names seem understandable just by reading them without any knowledge of the method.

**Pull request merged**